plugins/UM3NetworkPrinting/tests/TestSendMaterialJob.py

```diff
@@ -0,0 +1,210 @@
+    _LOCAL_MATERIAL_WHITE = {"type": "material", "status": "unknown", "id": "generic_pla_white",
+                             "base_file": "generic_pla_white", "setting_version": "5", "name": "White PLA",
+                             "brand": "Generic", "material": "PLA", "color_name": "White",
+                             "GUID": "badb0ee7-87c8-4f3f-9398-938587b67dce", "version": "1", "color_code": "#ffffff",
+                             "description": "Test PLA White", "adhesion_info": "Use glue.", "approximate_diameter": "3",
+                             "properties": {"density": "1.00", "diameter": "2.85", "weight": "750"},
+                             "definition": "fdmprinter", "compatible": True}
+
+    _LOCAL_MATERIAL_BLACK = {"type": "material", "status": "unknown", "id": "generic_pla_black",
+                             "base_file": "generic_pla_black", "setting_version": "5", "name": "Yellow CPE",
+                             "brand": "Ultimaker", "material": "CPE", "color_name": "Black",
+                             "GUID": "5fbb362a-41f9-4818-bb43-15ea6df34aa4", "version": "1", "color_code": "#000000",
+                             "description": "Test PLA Black", "adhesion_info": "Use glue.", "approximate_diameter": "3",
+                             "properties": {"density": "1.01", "diameter": "2.85", "weight": "750"},
+                             "definition": "fdmprinter", "compatible": True}
+
+    _REMOTE_MATERIAL_WHITE = {
+        "guid": "badb0ee7-87c8-4f3f-9398-938587b67dce",
+        "material": "PLA",
+        "brand": "Generic",
+        "version": 1,
+        "color": "White",
+        "density": 1.00
+    }
+
+    _REMOTE_MATERIAL_BLACK = {
+        "guid": "5fbb362a-41f9-4818-bb43-15ea6df34aa4",
+        "material": "PLA",
+        "brand": "Generic",
+        "version": 2,
+        "color": "Black",
+        "density": 1.00
+    }
+
+    @patch("plugins.UM3NetworkPrinting.src.ClusterUM3OutputDevice")
+    def test_run(self, device_mock):
+        job = SendMaterialJob(device_mock)
+        job.run()
+
+        # We expect the materials endpoint to be called when the job runs.
+        device_mock.get.assert_called_with("materials/", on_finished = job._onGetRemoteMaterials)
+
+    @patch("plugins.UM3NetworkPrinting.src.ClusterUM3OutputDevice")
+    @patch("PyQt5.QtNetwork.QNetworkReply")
+    def test__onGetRemoteMaterials_withFailedRequest(self, reply_mock, device_mock):
+        reply_mock.attribute.return_value = 404
+        job = SendMaterialJob(device_mock)
+        job._onGetRemoteMaterials(reply_mock)
+
+        # We expect the error string to be retrieved and the device not to be called for any follow up.
+        self.assertEqual([call.attribute(0), call.errorString()], reply_mock.method_calls)
+        self.assertEqual(0, device_mock.createFormPart.call_count)
+
+    @patch("plugins.UM3NetworkPrinting.src.ClusterUM3OutputDevice")
+    @patch("PyQt5.QtNetwork.QNetworkReply")
```

Comment:
It doesn't make sense to mock `QNetworkReply` because that could still mean the request is being actually sent via HTTP.
What we need is to mock the class that results in this reply, i.e. `postFormWithParts.mock_calls[0][2]['on_finished'](reply_mock)`

Reply:
I'm not sure what you mean.